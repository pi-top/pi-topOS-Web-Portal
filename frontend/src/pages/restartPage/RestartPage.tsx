import React, { useState, useEffect } from "react";
import { Line as ProgressBar } from "rc-progress";

import Layout from "../../components/layout/Layout";
import Spinner from "../../components/atoms/spinner/Spinner";

import rebootScreen from "../../assets/images/reboot-screen.png";
import styles from "./RestartPage.module.css";

import { runningOnWebRenderer } from "../../helpers/utils";
import ManualPowerOnDialogContainer from "./manualPowerOnDialog/ManualPowerOnDialogContainer";

export enum ErrorMessage {
  GlobalError = "Something went wrong while setting me up! Please click 'Restart' and contact support@pi-top.com if you experience any problems",
  RebootError = "I can't get to sleep! Please hold my power button down - that always makes me sleepy",
  TimeoutError = "It seems as though this is taking longer than expected. Try refreshing this webpage in your browser. Otherwise, try checking your connection and restarting your pi-top device."
}

export enum ServerStatusMessages {
  Waiting = "Rebooting. Please wait until this pi-top device is back online",
  Online = "The device is back online!"
}

export enum ExplanationMessages {
  OnWebRenderer = "Press 'restart' and I'll set some stuff up before rebooting. This might take up to a couple of minutes.",
  OnBrowser = "Press 'restart' to apply some final changes to your pi-top device and restart it.\n\nThis page will automatically update when the device is ready - this might take up to a couple of minutes, so don't go anywhere"
}

export type Props = {
  isWaitingForServer: boolean
  serverRebooted: boolean
  globalError: boolean;
  isSettingUpDevice: boolean;
  rebootError: boolean;
  progressPercentage: number;
  progressMessage: string;
  displayManualPowerOnDialog: boolean;
  onManualPowerOnDialogClose: () => void;
  setupDevice: () => void;
  onBackClick?: () => void;
};

export default ({
  isWaitingForServer,
  serverRebooted,
  globalError,
  isSettingUpDevice,
  rebootError,
  progressPercentage,
  progressMessage,
  displayManualPowerOnDialog,
  onManualPowerOnDialogClose,
  setupDevice,
  onBackClick,
}: Props) => {
  const [manualPowerOnDialogActive, setManualPowerOnDialogActive] = useState(false);

  useEffect(() => {
    setManualPowerOnDialogActive(displayManualPowerOnDialog);
  }, [displayManualPowerOnDialog])

  let errorMessage = "";
  if (globalError) {
    errorMessage = ErrorMessage.GlobalError;
  }

  if (rebootError) {
    errorMessage = isWaitingForServer ? ErrorMessage.TimeoutError : ErrorMessage.RebootError;
  }

  let explanationMessage = runningOnWebRenderer() ? ExplanationMessages.OnWebRenderer : ExplanationMessages.OnBrowser;

  return (
    <Layout
      banner={{
        src: rebootScreen,
        alt: "reboot-screen",
      }}
      prompt={
        <>
          Right, I need a quick <span className="green">nap</span>
        </>
      }
      explanation={explanationMessage}
      nextButton={{
        onClick: setupDevice,
        label: "Restart",
        disabled: isSettingUpDevice || rebootError || isWaitingForServer,
        hidden: isWaitingForServer || serverRebooted
      }}
      backButton={
        (globalError || !onBackClick)
          ? undefined
          : {
              onClick: onBackClick,
              disabled: isSettingUpDevice || isWaitingForServer,
              hidden: isWaitingForServer || serverRebooted
            }
      }
      className={styles.root}
    >

      <ManualPowerOnDialogContainer
        active={manualPowerOnDialogActive}
        onClose={() => {
          setManualPowerOnDialogActive(!manualPowerOnDialogActive);
          onManualPowerOnDialogClose();
        }}
      />

      {isSettingUpDevice && (
        <div className={styles.progress}>
          <ProgressBar
            percent={progressPercentage}
            strokeWidth={2}
            strokeColor="#71c0b4"
          />
          <span className={styles.message}>{progressMessage}</span>
        </div>
      )}
      {isWaitingForServer ? (
        <>
          <span className={styles.message}>{ServerStatusMessages.Waiting}</span>
          <Spinner size={40} />{" "}
        </>
      ):(
        serverRebooted && <span className={styles.message}>{ServerStatusMessages.Online}</span>
      )}

      {errorMessage && <span className={styles.error}>{errorMessage}</span>}
    </Layout>
  );
};
