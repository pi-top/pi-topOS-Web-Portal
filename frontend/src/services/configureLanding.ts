import axios from "axios";

import apiBaseUrl from "./apiBaseUrl";

export default async function configureLanding() {
  await axios.post(
    `${apiBaseUrl}/configure-landing`,
    {}
  );
}
