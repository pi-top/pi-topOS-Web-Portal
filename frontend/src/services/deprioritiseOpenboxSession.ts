import api from "./api";

export default async function deprioritiseOpenboxSession() {
  await api.post(`/deprioritise-openbox-session`, {});
}
