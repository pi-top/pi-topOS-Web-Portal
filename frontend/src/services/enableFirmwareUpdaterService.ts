import api from "./api";

export default async function enableFirmwareUpdaterService() {
  await api.post(`/enable-firmware-updater-service`, {});
}
