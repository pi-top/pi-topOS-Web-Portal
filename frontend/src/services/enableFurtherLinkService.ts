import api from "./api";

export default async function enableFurtherLinkService() {
  await api.post(`/enable-further-link-service`, {});
}
