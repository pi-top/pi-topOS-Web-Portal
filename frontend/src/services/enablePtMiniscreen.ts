import api from "./api";

export default async function enablePtMiniscreen() {
  await api.post(`/enable-pt-miniscreen`, {});
}
