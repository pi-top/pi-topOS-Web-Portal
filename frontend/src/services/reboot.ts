import axios from "axios";

import apiBaseUrl from "./apiBaseUrl";

export default async function reboot() {
  await axios.post(
    `${apiBaseUrl}/reboot`,
    {}
  );
}
