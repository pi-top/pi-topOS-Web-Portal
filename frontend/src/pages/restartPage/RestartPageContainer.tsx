import React, { useState, useEffect } from "react";
import { useHistory } from 'react-router-dom';

import semver from 'semver';
import RestartPage from "./RestartPage";

import configureLanding from "../../services/configureLanding";
import deprioritiseOpenboxSession from "../../services/deprioritiseOpenboxSession";
import enableFirmwareUpdaterService from "../../services/enableFirmwareUpdaterService";
import enableFurtherLinkService from "../../services/enableFurtherLinkService";
import stopOnboardingAutostart from "../../services/stopOnboardingAutostart";
import reboot from "../../services/reboot";
import restoreFiles from "../../services/restoreFiles";
import serverStatus from "../../services/serverStatus"
import updateEeprom from "../../services/updateEeprom"
import enablePtMiniscreen from "../../services/enablePtMiniscreen";
import updateHubFirmware from "../../services/updateHubFirmware";
import getBuildInfo from "../../services/getBuildInfo";

import { runningOnWebRenderer } from "../../helpers/utils";

const maxProgress = 11; // this is the number of services for setting up

const calculatePercentageProgress = (progress: number, maxProgress: number) => {
  if (!Number.isFinite(progress / maxProgress)) {
    return 100;
  }

  return Math.floor((progress / maxProgress) * 100);
};

export type Props = {
  globalError?: boolean;
  goToPreviousPage?: () => void;
};

export default ({
  globalError = false,
  goToPreviousPage,
}: Props) => {
  const history = useHistory();
  const [isSettingUpDevice, setIsSettingUpDevice] = useState(false);
  const [rebootError, setRebootError] = useState(false);
  const [progressMessage, setProgressMessage] = useState("Alright let's get started!");
  const [progress, setProgress] = useState(0);
  const [isWaitingForServer, setIsWaitingForServer] = useState(false);
  const [serverRebooted, setServerRebooted] = useState(false);
  const [manualPowerOnRequired, setManualPowerOnRequired] = useState(false);
  const [displayManualPowerOnDialog, setDisplayManualPowerOnDialog] = useState(false);

  useEffect(() => {
    getBuildInfo()
      .then((buildInfo) => {
        setManualPowerOnRequired(semver.lt(buildInfo.hubFirmwareVersion, "3.0"));
      })
      .catch(() => setManualPowerOnRequired(false))
  }, [])

  function safelyRunService(service: () => Promise<void>, message: string) {
    return service()
      .then(() => setProgressMessage(message))
      .catch(() =>
        setProgressMessage(
          "I couldn't do that, please contact support if I have any problems..."
        )
      )
      .finally(() => {
        setProgress(currentProgess =>
          Math.min(currentProgess + 1, maxProgress)
        );
      });
  }

  const rebootTimeoutMs = 120000;
  const timeoutServerStatusRequestMs = 500;
  const serverStatusRequestIntervalMs = 1500;
  let elapsedWaitingTimeMs = 0;

  function waitUntilServerIsOnline() {
    const interval = setInterval(async () => {
      try {
        elapsedWaitingTimeMs += timeoutServerStatusRequestMs + serverStatusRequestIntervalMs;
        elapsedWaitingTimeMs >= rebootTimeoutMs && setRebootError(true);
        await serverStatus({ timeout: timeoutServerStatusRequestMs });
        setServerRebooted(true);
        setIsWaitingForServer(false);
        clearInterval(interval);
        history.push("/");
        window.location.reload()
      } catch (_) {}
    }, serverStatusRequestIntervalMs);
  }

  function rebootPiTop() {
    reboot()
      .then(() => {
        if (!runningOnWebRenderer()) {
          setIsSettingUpDevice(false);
          setIsWaitingForServer(true);
          setServerRebooted(false);
          window.setTimeout(waitUntilServerIsOnline, 3000);
        }
      })
      .catch(() => {
        setRebootError(true);
        setIsSettingUpDevice(false);
      })
  }


  return (
    <RestartPage
      displayManualPowerOnDialog={displayManualPowerOnDialog}
      onManualPowerOnDialogClose={rebootPiTop}
      isWaitingForServer={isWaitingForServer}
      serverRebooted={serverRebooted}
      globalError={globalError}
      isSettingUpDevice={isSettingUpDevice}
      rebootError={rebootError}
      progressPercentage={calculatePercentageProgress(progress, maxProgress)}
      progressMessage={progressMessage}
      onBackClick={goToPreviousPage}
      setupDevice={() => {
        setIsSettingUpDevice(true);

        safelyRunService(
          enableFirmwareUpdaterService,
          "Reminded myself to keep an eye out for cool new stuff for my friends..."
        )
          .finally(() =>
            safelyRunService(
              enableFurtherLinkService,
              "Reminded myself to stay connected, so I can help you go Further..."
            )
          )
          .finally(() =>
            safelyRunService(
              deprioritiseOpenboxSession,
              "Put away all my open boxes..."
            )
          )
          .finally(() =>
            safelyRunService(
              restoreFiles,
              "Reminded myself to show you around..."
            )
          )
          .finally(() =>
            safelyRunService(
              configureLanding,
              "Reminded myself to show you around..."
            )
          )
          .finally(() =>
            safelyRunService(
              stopOnboardingAutostart,
              "Made sure not to make you go through this again..."
            )
          )
          .finally(() =>
            safelyRunService(
              updateHubFirmware,
              "Made things easier for me to go to sleep when you ask..."
            )
          )
          .finally(() =>
            safelyRunService(
              updateEeprom,
              "Made things easier for me to go to sleep when you ask..."
            )
          )
          .finally(() =>
            safelyRunService(
              enablePtMiniscreen,
              "Reminded myself to tell the miniscreen to do its thing in the morning..."
            )
          )
          .catch(console.error)
          .finally(() => {
            if (manualPowerOnRequired) {
              setDisplayManualPowerOnDialog(true);
            } else {
              rebootPiTop();
            }
          })
      }}
    />
  );
};
