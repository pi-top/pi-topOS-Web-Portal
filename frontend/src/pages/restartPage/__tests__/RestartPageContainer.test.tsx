import React from "react";
import {
  render,
  BoundFunction,
  QueryByText,
  GetByText,
  RenderResult,
  wait,
  fireEvent,
  getByLabelText,
  waitForElement,
} from "@testing-library/react";

import RestartPageContainer, { Props } from "../RestartPageContainer";
import { ErrorMessage, ExplanationMessages } from "../RestartPage";
import querySpinner from "../../../../test/helpers/querySpinner";

import configureLanding from "../../../services/configureLanding";
import deprioritiseOpenboxSession from "../../../services/deprioritiseOpenboxSession";
import enableFurtherLinkService from "../../../services/enableFurtherLinkService";
import enableFirmwareUpdaterService from "../../../services/enableFirmwareUpdaterService";
import restoreFiles from "../../../services/restoreFiles";
import stopOnboardingAutostart from "../../../services/stopOnboardingAutostart";
import reboot from "../../../services/reboot";
import serverStatus from "../../../services/serverStatus";
import updateEeprom from "../../../services/updateEeprom";
import enablePtMiniscreen from "../../../services/enablePtMiniscreen";
import updateHubFirmware from "../../../services/updateHubFirmware";

import { act } from "react-dom/test-utils";

jest.mock("../../../services/configureLanding");
jest.mock("../../../services/deprioritiseOpenboxSession");
jest.mock("../../../services/enableFurtherLinkService");
jest.mock("../../../services/enableFirmwareUpdaterService");
jest.mock("../../../services/restoreFiles");
jest.mock("../../../services/stopOnboardingAutostart");
jest.mock("../../../services/reboot");
jest.mock("../../../services/serverStatus");
jest.mock("../../../services/updateEeprom");
jest.mock("../../../services/enablePtMiniscreen");
jest.mock("../../../services/updateHubFirmware");


const configureLandingMock = configureLanding as jest.Mock;
const deprioritiseOpenboxSessionMock = deprioritiseOpenboxSession as jest.Mock;
const enableFurtherLinkServiceMock = enableFurtherLinkService as jest.Mock;
const enableFirmwareUpdaterServiceMock = enableFirmwareUpdaterService as jest.Mock;
const restoreFilesMock = restoreFiles as jest.Mock;
const stopOnboardingAutostartMock = stopOnboardingAutostart as jest.Mock;
const rebootMock = reboot as jest.Mock;
const serverStatusMock = serverStatus as jest.Mock;
const updateEepromMock = updateEeprom as jest.Mock;
const enablePtMiniscreenMock = enablePtMiniscreen as jest.Mock;
const updateHubFirmwareMock = updateHubFirmware as jest.Mock;


const mockServices = [
  configureLandingMock,
  deprioritiseOpenboxSessionMock,
  enableFurtherLinkServiceMock,
  enableFirmwareUpdaterServiceMock,
  restoreFilesMock,
  stopOnboardingAutostartMock,
  updateEepromMock,
  updateHubFirmwareMock,
  rebootMock,
  enablePtMiniscreenMock,
];


const resolveMocks = (mocks: jest.Mock[] = mockServices) => {
  mocks.forEach((mock) => {
    mock.mockResolvedValue("OK");
  });
};

const rejectMocks = (mocks: jest.Mock[] = mockServices) => {
  mocks.forEach((mock, i) => {
    mock.mockRejectedValue(new Error(`Test error message ${i}`));
  });
};

let mockUserAgent = "web-renderer";
Object.defineProperty(global.navigator, "userAgent", {
  get() {
    return mockUserAgent;
  },
});


jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    push: jest.fn(),
  }),
}));

const userMustConfirmBeforeLeaving = () => {
  // dispatch beforeunload event to simulate beginning of unload process
  const event = new Event('beforeunload')
  const preventDefaultSpy = jest.spyOn(event, "preventDefault");
  window.dispatchEvent(event);

  // calling preventDefault causes most browsers to confirm leaving with user
  // event.returnValue needs to be set for chrome
  return (
    preventDefaultSpy.mock.calls.length > 0 &&
    event.returnValue
  );
};

describe("RestartPageContainer", () => {
  let defaultProps: Props;
  let restartPageContainer: HTMLElement;
  let queryByText: BoundFunction<QueryByText>;
  let getByText: BoundFunction<GetByText>;
  let rerender: RenderResult["rerender"];
  beforeEach(async () => {
    resolveMocks();
    serverStatusMock.mockResolvedValue("OK");
    mockUserAgent = "web-renderer";
    defaultProps = {};

    ({
      container: restartPageContainer,
      queryByText,
      getByText,
      rerender,
    } = render(<RestartPageContainer {...defaultProps} />));
  });
  afterEach(() => {
    mockServices.forEach((mock) => {
      mock.mockRestore();
    });
  });

  it("does not render back button", () => {
    expect(queryByText("Back")).not.toBeInTheDocument();
  });

  it("does not render error message", () => {
    expect(
      restartPageContainer.querySelector(".error")
    ).not.toBeInTheDocument();
  });

  it("does not ask user for confirmation before they leave the page", async () => {
    expect(userMustConfirmBeforeLeaving()).toBeFalsy();
  });

  describe("when globalError is true", () => {
    beforeEach(() => {
      defaultProps = {
        ...defaultProps,
        globalError: true,
      };

      rerender(<RestartPageContainer {...defaultProps} />);
    });

    it("renders correct error message", () => {
      expect(getByText(ErrorMessage.GlobalError)).toBeInTheDocument();
    });
  });

  describe("when goToPreviousPage is passed", () => {
    beforeEach(() => {
      defaultProps = {
        ...defaultProps,
        goToPreviousPage: jest.fn(),
      };

      rerender(<RestartPageContainer {...defaultProps} />);
    });

    it("renders back button", () => {
      expect(queryByText("Back")).toBeInTheDocument();
    });

    it("calls goToPreviousPage on back button click", () => {
      fireEvent.click(getByText("Back"));

      expect(defaultProps.goToPreviousPage).toHaveBeenCalled();
    });

    it("disables back button on restart click", async () => {
      fireEvent.click(getByText("Restart"));

      expect(queryByText("Back")).toBeDisabled();

      await wait();
    });
  });

  describe("on restart click", () => {
    it("renders progress", async () => {
      fireEvent.click(getByText("Restart"));

      expect(
        restartPageContainer.querySelector(".progress")
      ).toBeInTheDocument();

      await wait();
    });

    it("disables restart button", async () => {
      fireEvent.click(getByText("Restart"));

      expect(queryByText("Restart")).toBeDisabled();

      await wait();
    });

    it("asks user for confirmation before they leave the page", async () => {
      fireEvent.click(getByText("Restart"));

      await wait();

      expect(userMustConfirmBeforeLeaving()).toBeTruthy();

      await wait();
    });

    it("calls correct services", async () => {
      fireEvent.click(getByText("Restart"));

      await wait();

      mockServices.forEach((mock) => {
        expect(mock).toHaveBeenCalled();
      });
    });

    describe('when configure landing fails', () => {
      beforeEach(() => {
        configureLandingMock.mockRejectedValue(new Error());
      });

      it('calls remaining services', async () => {
        fireEvent.click(getByText("Restart"));

        await wait();

        mockServices.forEach((mock) => {
          expect(mock).toHaveBeenCalled();
        });
      });
    });

    describe('when deprioritise openbox session fails', () => {
      beforeEach(() => {
        deprioritiseOpenboxSessionMock.mockRejectedValue(new Error());
      });

      it('calls remaining services', async () => {
        fireEvent.click(getByText("Restart"));

        await wait();

        mockServices.forEach((mock) => {
          expect(mock).toHaveBeenCalled();
        });
      });
    });

    describe('when stop onboarding autostart fails', () => {
      beforeEach(() => {
        stopOnboardingAutostartMock.mockRejectedValue(new Error());
      });

      it('calls remaining services', async () => {
        fireEvent.click(getByText("Restart"));

        await wait();

        mockServices.forEach((mock) => {
          expect(mock).toHaveBeenCalled();
        });
      });
    });

    describe('when update EEPROM fails', () => {
      beforeEach(() => {
        updateEepromMock.mockRejectedValue(new Error());
      });

      it('calls remaining services', async () => {
        fireEvent.click(getByText("Restart"));

        await wait();

        mockServices.forEach((mock) => {
          expect(mock).toHaveBeenCalled();
        });
      });
    });

    describe('when update hub firmware fails', () => {
      beforeEach(() => {
        updateHubFirmwareMock.mockRejectedValue(new Error());
      });

      it('calls remaining services', async () => {
        fireEvent.click(getByText("Restart"));

        await wait();

        mockServices.forEach((mock) => {
          expect(mock).toHaveBeenCalled();
        });
      });
    });

    describe('when reboot fails', () => {
      beforeEach(async () => {
        rebootMock.mockRejectedValue(new Error());

        fireEvent.click(getByText("Restart"));

        await wait();
      });

      it('disables restart button', async () => {
        expect(queryByText('Restart')).toBeDisabled();
      });

      it('renders reboot error message', () => {
        expect(queryByText(ErrorMessage.RebootError)).toBeInTheDocument();
      });

      it('stops rendering progress', () => {
        expect(restartPageContainer.querySelector('.progress')).not.toBeInTheDocument();
      })
    });

    describe('after reboot call when running through an external browser', () => {
      beforeEach(async () => {
        jest.useFakeTimers();
        mockUserAgent = "not-web-renderer";
        serverStatusMock.mockResolvedValue("OK");
        fireEvent.click(getByText("Restart"));
        await wait();
      });
      afterEach(() => jest.useRealTimers());

      it('displays a wait message', async () => {
        ExplanationMessages.OnBrowser.split('\n').map(function(item, _) {
          item && expect(getByText(item)).toBeInTheDocument();
        });

        await act(async () => {
          jest.runOnlyPendingTimers();
          await Promise.resolve();
        });
      });

      it('starts querying device to learn if its back online', async () => {
        await act(async () => {
          jest.runOnlyPendingTimers();
          await Promise.resolve();
        });
        expect(serverStatusMock).toHaveBeenCalled();
      });

      it('renders a spinner', async () => {
        await act(async () => {
          jest.runOnlyPendingTimers();
          await Promise.resolve();
        });
        expect(querySpinner(restartPageContainer)).toBeInTheDocument();
      });

      it("asks user for confirmation before they leave the page", async () => {
        fireEvent.click(getByText("Restart"));

        await wait();

        expect(userMustConfirmBeforeLeaving()).toBeTruthy();

        await wait();
      });

      describe('when the device is back online', () => {
        it('updates the displayed message', async () => {
          await act(async () => {
            jest.runOnlyPendingTimers();
            jest.runOnlyPendingTimers();
            await Promise.resolve();
          });
          expect(getByText("The device is back online!")).toBeInTheDocument()
        });

        it('doesn\'t render a spinner', async () => {
          await act(async () => {
            jest.runOnlyPendingTimers();
            jest.runOnlyPendingTimers();
            await Promise.resolve();
          });
          expect(querySpinner(restartPageContainer)).not.toBeInTheDocument();
        });

        it("does not ask user for confirmation before they leave the page", async () => {
          await act(async () => {
            jest.runOnlyPendingTimers();
            jest.runOnlyPendingTimers();
            await Promise.resolve();
          });
          expect(userMustConfirmBeforeLeaving()).toBeFalsy();
        });
      });
    });
  });
});
