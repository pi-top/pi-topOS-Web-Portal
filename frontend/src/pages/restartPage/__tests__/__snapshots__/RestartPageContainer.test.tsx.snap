// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RestartPageContainer renders prompt correctly 1`] = `
<h1
  class="prompt"
>
  Right, I need a quick 
  <span
    class="green"
  >
    nap
  </span>
</h1>
`;
